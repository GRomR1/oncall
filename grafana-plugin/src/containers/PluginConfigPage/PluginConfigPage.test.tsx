import React from 'react';

import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer, { act, ReactTestRenderer } from 'react-test-renderer';
import { OnCallPluginConfigPageProps } from 'types';

import {
  checkIfPluginIsConnected as checkIfPluginIsConnectedOriginal,
  resetPlugin as resetPluginOriginal,
  syncDataWithOnCall as syncDataWithOnCallOriginal,
  selfHostedInstallPlugin as selfHostedInstallPluginOriginal,
} from 'state/plugin';

import PluginConfigPage from './PluginConfigPage';

const checkIfPluginIsConnected = checkIfPluginIsConnectedOriginal as jest.Mock<
  ReturnType<typeof checkIfPluginIsConnectedOriginal>
>;
const resetPlugin = resetPluginOriginal as jest.Mock<ReturnType<typeof resetPluginOriginal>>;
const selfHostedInstallPlugin = selfHostedInstallPluginOriginal as jest.Mock<
  ReturnType<typeof selfHostedInstallPluginOriginal>
>;
const syncDataWithOnCall = syncDataWithOnCallOriginal as jest.Mock<ReturnType<typeof syncDataWithOnCallOriginal>>;

jest.mock('state/plugin');

enum License {
  OSS = 'OpenSource',
  CLOUD = 'some-other-license',
}

const SELF_HOSTED_INSTALL_PLUGIN_ERROR_MESSAGE = 'ohhh nooo an error msg from self hosted install plugin';
const CHECK_IF_PLUGIN_IS_CONNECTED_ERROR_MESSAGE = 'ohhh nooo a plugin connection error';
const SNYC_DATA_WITH_ONCALL_ERROR_MESSAGE = 'ohhh noooo a sync issue';

const mockSyncDataWithOnCall = (license: License = License.OSS) => {
  syncDataWithOnCall.mockResolvedValueOnce({
    token_ok: true,
    license,
    version: 'v1.2.3',
  });
};

const generateComponentProps = (
  onCallApiUrl: OnCallPluginConfigPageProps['plugin']['meta']['jsonData']['onCallApiUrl'] = null
): OnCallPluginConfigPageProps =>
  ({
    plugin: {
      meta: {
        jsonData: onCallApiUrl === null ? null : { onCallApiUrl },
      },
    },
  } as OnCallPluginConfigPageProps);

describe('ConfigurationForm', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("If onCallApiUrl is not set in the plugin's meta jsonData, or in process.env, checkIfPluginIsConnected is not called, and the configuration form is shown", async () => {
    // mocks
    delete process.env.ONCALL_API_URL;

    checkIfPluginIsConnected.mockImplementation();
    syncDataWithOnCall.mockImplementation();

    // test setup
    let component: ReactTestRenderer;
    await act(async () => {
      component = renderer.create(<PluginConfigPage {...generateComponentProps()} />);
    });

    // assertions
    expect(checkIfPluginIsConnected).not.toHaveBeenCalled();
    expect(syncDataWithOnCall).not.toHaveBeenCalled();
    expect(component).toMatchSnapshot();
  });

  test("If onCallApiUrl is not set in the plugin's meta jsonData, and ONCALL_API_URL is passed in process.env, it calls selfHostedInstallPlugin", async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    mockSyncDataWithOnCall();

    // test setup
    await act(async () => {
      renderer.create(<PluginConfigPage {...generateComponentProps()} />);
    });

    // assertions
    expect(selfHostedInstallPlugin).toHaveBeenCalledTimes(1);
    expect(selfHostedInstallPlugin).toHaveBeenCalledWith(processEnvOnCallApiUrl);
  });

  test("If onCallApiUrl is not set in the plugin's meta jsonData, and ONCALL_API_URL is passed in process.env, and there is an error calling selfHostedInstallPlugin, it sets an error message", async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    selfHostedInstallPlugin.mockResolvedValueOnce(SELF_HOSTED_INSTALL_PLUGIN_ERROR_MESSAGE);

    // test setup
    let component: ReactTestRenderer;
    await act(async () => {
      component = renderer.create(<PluginConfigPage {...generateComponentProps()} />);
    });

    // assertions
    expect(selfHostedInstallPlugin).toHaveBeenCalledTimes(1);
    expect(selfHostedInstallPlugin).toHaveBeenCalledWith(processEnvOnCallApiUrl);
    expect(component).toMatchSnapshot();
  });

  test('If onCallApiUrl is set, and checkIfPluginIsConnected returns an error, it sets an error message', async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    checkIfPluginIsConnected.mockResolvedValueOnce(CHECK_IF_PLUGIN_IS_CONNECTED_ERROR_MESSAGE);

    // test setup
    let component: ReactTestRenderer;
    await act(async () => {
      component = renderer.create(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    });

    // assertions
    expect(checkIfPluginIsConnected).toHaveBeenCalledTimes(1);
    expect(checkIfPluginIsConnected).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
    expect(component).toMatchSnapshot();
  });

  test('OnCallApiUrl is set, and syncDataWithOnCall returns an error', async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    checkIfPluginIsConnected.mockResolvedValueOnce(null);
    syncDataWithOnCall.mockResolvedValueOnce(SNYC_DATA_WITH_ONCALL_ERROR_MESSAGE);

    // test setup
    let component: ReactTestRenderer;
    await act(async () => {
      component = renderer.create(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    });

    // assertions
    expect(checkIfPluginIsConnected).toHaveBeenCalledTimes(1);
    expect(checkIfPluginIsConnected).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
    expect(component).toMatchSnapshot();
  });

  test.each([License.CLOUD, License.OSS])(
    'OnCallApiUrl is set, and syncDataWithOnCall does not return an error. It displays properly the plugin connected items based on the license - License: %s',
    async (license) => {
      // mocks
      const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
      const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

      process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

      checkIfPluginIsConnected.mockResolvedValueOnce(null);
      mockSyncDataWithOnCall(license);

      // test setup
      let component: ReactTestRenderer;
      await act(async () => {
        component = renderer.create(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
      });

      // assertions
      expect(checkIfPluginIsConnected).toHaveBeenCalledTimes(1);
      expect(checkIfPluginIsConnected).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
      expect(component).toMatchSnapshot();
    }
  );

  // TODO:
  test.skip('Successful plugin reset', async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    checkIfPluginIsConnected.mockResolvedValueOnce(null);
    mockSyncDataWithOnCall(License.OSS);
    resetPlugin.mockResolvedValueOnce();

    // test setup
    let component: RenderResult;
    await act(async () => {
      component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    });

    // // TODO:
    // // const user = userEvent.setup();
    // // await user.click(screen.getByRole('button'));

    const user = userEvent.setup();
    // const component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    await user.click(screen.getByRole('button'));

    // assertions
    expect(checkIfPluginIsConnected).toHaveBeenCalledTimes(1);
    expect(checkIfPluginIsConnected).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);

    expect(syncDataWithOnCall).toHaveBeenCalledTimes(1);
    expect(syncDataWithOnCall).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);

    expect(resetPlugin).toHaveBeenCalledTimes(1);
    expect(resetPlugin).toHaveBeenCalledWith();

    expect(component).toMatchSnapshot();
  });

  // TODO:
  test.skip('Unsuccessful plugin reset', () => {});
});
