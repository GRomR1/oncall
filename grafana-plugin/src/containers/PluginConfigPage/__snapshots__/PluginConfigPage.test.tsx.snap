// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationForm If onCallApiUrl is not set in the plugin's meta jsonData, and ONCALL_API_URL is passed in process.env, and there is an error calling selfHostedInstallPlugin, it sets an error message 1`] = `
[
  <legend
    className="css-xim8hk"
  >
    Configure Grafana OnCall
  </legend>,
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>,
  <pre>
    <span
      className="root text text--undefined text--medium"
    >
      ohhh nooo an error msg from self hosted install plugin
    </span>
  </pre>,
  <button
    className="css-mk7eo3-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="css-1mhnkuh"
    >
      Remove current configuration
    </span>
  </button>,
]
`;

exports[`ConfigurationForm If onCallApiUrl is not set in the plugin's meta jsonData, or in process.env, checkIfPluginIsConnected is not called, and the configuration form is shown 1`] = `
[
  <legend
    className="css-xim8hk"
  >
    Configure Grafana OnCall
  </legend>,
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>,
  <form
    className="css-xs0vux"
    onSubmit={[Function]}
  >
    <div
      className="info-block"
    >
      <p>
        1. Launch the OnCall backend
      </p>
      <span
        className="root text text--secondary text--medium"
      >
        Run hobby, dev or production backend. See
         
        <a
          href="https://github.com/grafana/oncall#getting-started"
          rel="noreferrer"
          target="_blank"
        >
          <span
            className="root text text--link text--medium"
          >
            here
          </span>
        </a>
         
        on how to get started.
      </span>
    </div>
    <div
      className="info-block"
    >
      <p>
        2. Let us know the base URL of your OnCall API
      </p>
      <span
        className="root text text--secondary text--medium"
      >
        The OnCall backend must be reachable from your Grafana installation. Some examples are:
        <br />
        - http://host.docker.internal:8080
        <br />
        - http://localhost:8080
      </span>
    </div>
    <div
      className="css-8e5b3"
    >
      <div
        className="css-jt4xma-Label"
      >
        <label>
          <div
            className="css-xhqy0o"
          >
            OnCall backend URL
          </div>
        </label>
      </div>
      <div>
        <div
          className="css-xcstkt-input-wrapper"
          data-testid="input-wrapper"
        >
          <div
            className="css-1w5c5dq-input-inputWrapper"
          >
            <input
              className="css-1wdli31-input-input"
              data-testid="onCallApiUrl"
              name="onCallApiUrl"
              onBlur={[Function]}
              onChange={[Function]}
              style={
                {
                  "paddingLeft": undefined,
                  "paddingRight": undefined,
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
    <button
      className="css-1sara2j-button"
      disabled={false}
      type="submit"
    >
      <span
        className="css-1mhnkuh"
      >
        Connect
      </span>
    </button>
  </form>,
]
`;

exports[`ConfigurationForm If onCallApiUrl is set, and checkIfPluginIsConnected returns an error, it sets an error message 1`] = `
[
  <legend
    className="css-xim8hk"
  >
    Configure Grafana OnCall
  </legend>,
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>,
  <pre>
    <span
      className="root text text--undefined text--medium"
    >
      ohhh nooo a plugin connection error
    </span>
  </pre>,
  <button
    className="css-mk7eo3-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="css-1mhnkuh"
    >
      Remove current configuration
    </span>
  </button>,
]
`;

exports[`ConfigurationForm OnCallApiUrl is set, and syncDataWithOnCall does not return an error. It displays properly the plugin connected items based on the license - License: OpenSource 1`] = `
[
  <legend
    className="css-xim8hk"
  >
    Configure Grafana OnCall
  </legend>,
  <p>
    Plugin is connected! Check Grafana OnCall ðŸ‘ˆðŸ‘ˆðŸ‘ˆ 
    <img
      alt="Grafana OnCall Logo"
      src={
        {
          "getCacheKey": [Function],
          "process": [Function],
        }
      }
      width={18}
    />
  </p>,
  <pre>
    <span
      className="root text text--undefined text--medium"
    >
      Connected to OnCall (v1.2.3, OpenSource)
    </span>
  </pre>,
  <button
    className="css-mk7eo3-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="css-1mhnkuh"
    >
      Remove current configuration
    </span>
  </button>,
]
`;

exports[`ConfigurationForm OnCallApiUrl is set, and syncDataWithOnCall does not return an error. It displays properly the plugin connected items based on the license - License: some-other-license 1`] = `
[
  <legend
    className="css-xim8hk"
  >
    Configure Grafana OnCall
  </legend>,
  <p>
    Plugin is connected! Check Grafana OnCall ðŸ‘ˆðŸ‘ˆðŸ‘ˆ 
    <img
      alt="Grafana OnCall Logo"
      src={
        {
          "getCacheKey": [Function],
          "process": [Function],
        }
      }
      width={18}
    />
  </p>,
  <pre>
    <span
      className="root text text--undefined text--medium"
    >
      Connected to OnCall (v1.2.3, some-other-license)
    </span>
  </pre>,
  <div
    className="css-jt4xma-Label"
  >
    <label>
      <div
        className="css-xhqy0o"
      >
        This is a cloud managed configuration.
      </div>
    </label>
  </div>,
]
`;

exports[`ConfigurationForm OnCallApiUrl is set, and syncDataWithOnCall returns an error 1`] = `
[
  <legend
    className="css-xim8hk"
  >
    Configure Grafana OnCall
  </legend>,
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>,
  <pre>
    <span
      className="root text text--undefined text--medium"
    >
      ohhh noooo a sync issue
    </span>
  </pre>,
  <button
    className="css-1sara2j-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="css-1mhnkuh"
    >
      Retry Sync
    </span>
  </button>,
]
`;
